### ISSUE TITLE
Develop more flexible form controls

### ISSUE THREAD

**Issue by viktor-yakubiv** [oacore/design#197]
@Joozty I would like to discuss the following interface:

```jsx
<Form.Group>
  <Form.Label>User name</Form.Label>
  <Form.Control value="john" placeholer="e.g. johndoe" />
  <Form.Control.PrependIcon src="user" />
  <Form.Text>Nice username</Form.Text>
</Form.Group>
```

It's a wrapper around HTML. However, it applies classes and can possibly sort elements, i.e. put the label in the end, after the input.

The TextField component will remain but would be a shortcut for this expanded group.

**Comment by Joozty** [comment]
I am quite neutral here. I understand we need more control for inputs, on the contrary, we would need 6 lines of code to develop simple input. I like TextFields for their simplicity. So inclining more to use this only internally and the user should use it only in edge cases.

All the props and children in your example are just strings. I doubt it can be anything else. So it's the same as using just new props for TextField.


```
<TextField
 label="User name"
 value="john"
 placeholder="e.g. johndoe"
 icon="#success"
 text="Nice username"
/>
```

PS: My example has 7 lines of code. It's even worse 🤦 It was not a good idea to measure code lines. 😄

**Comment by viktor-yakubiv** [comment]
I still would prefer your example if I use the library somewhere. I have only one remark to your code: there may be more than a single icon and the interface is not clear enough yet.

I see 2 options:
- `icon` is the left any icon, `status: ['check', 'alert', 'loading']` is the right icon with the limited options
- `prependIcon` for the left one, `appendIcon` for the right one, both can be any

However, this issue is only about the internal flexible interface to easily adjust our styles to selects and maybe other fields and provide some interface for edge cases as you said.

Do you have any suggestions to the interface itself? I do not like this bit:

```jsx
 <Form.Control.PrependIcon src="user" />
```

**Comment by Joozty** [comment]
Do we really want to have two icons? If so then your first opinion looks better for me.

I think you picked a reasonable naming. I would maybe just call Form.Text -> Form.Description and rename Form -> Input but that's like really opinionated and minor...

**Comment by viktor-yakubiv** [comment]
I will start developing it like it was designed originally and then we will see what naming would be the best. Thank you for the feedback 👍

**Comment by viktor-yakubiv** [comment]
@Joozty do you have any libraries for the inspiration you like? I checked [ant.design](http://ant.design), I do not really like the API but it's not bad.

I took my ideas from the Reactstrap but it misses cases about the icons. Do you have any library that does it?

**Comment by Joozty** [comment]
Not really. I used [material-ui](https://material-ui.com/) for my school projects and I liked the API (except use styles components). They allow to have icons on each side and they pass it down as a prop.

See example: https://codesandbox.io/s/material-demo-t3vgx?file=/demo.js

Btw. they call it adornment which is a maybe better name for that subcomponent instead of Append/Prepend Icon.
